### DISCUSSION TITLE
Initialize all configs to validate requirements on app start.

### DISCUSSION THREAD

**Discussion by jhirn** [palkan/anyway_config#121] [General]
Ideally, i'd like my application to fail to start/deploy if required configuration is not in place. The `required` keyword and `on_load` blocks have gotten me quite far, but only fail on instantiation or when the config is referenced. Most of the configs are referenced in initializers, but not in all cases so I had a couple slip by. 

Short of doing iterating and instantiating every class in `config/configs`, is there a better way to do this?

**Comment by palkan** [comment]
A quick idea that comes into my mind is to instantiate an instance right where the class is defined:

```ruby
class MyConfig
  # ...
  require :x, :foo
end.new #=> here we try to initialize an instance right away
```

**Comment by jhirn** [comment]
That syntax looks like it should work but was not letting me call `.new` on the class. I was also hoping for something that was enforced and could not be forgotten/omitted by mistake. I ended up with this loop in application.rb

```ruby
# Eager instantiate configs to fail on startup if misconfigured
Dir["config/configs/*.rb"].each do |f|
  require_dependency("#{Dir.pwd}/#{f}")
end
ApplicationConfig.subclasses.each(&:new)
```

I'd really like to use the value assigned to `config.anyway_config.default_config_path = Rails.root.join("config", "configs")` but it seems to convert to a proc which accepts an argument and returns the yaml file. I can certainly work around this but assigning `Rails.root.join("config", "configs")` to a var or doing something like `Anyway::Settings.default_config_path.call("path").gsub(/path.yml$/, "*.rb")`. It's a little specific to my use case to suggest as a feature. But maybe this conversation provide ideas to others.

**Reply by jhirn** [reply]
Quick update. The above approach proved to be problematic when using `bin/rails credentials:[diff|edit|enroll]`. It seems the secrets are not made available to Anyway plus could introduce a chicken and egg problem. I can introduce something like `ApplicationConfig.subclasses.each(&:new) unless ENV["SKIP_CONFIG_EAGER_LOAD"] == "true"` for the editing but the `.gitattributes` diff command still didn't seem to respect it. 🤷‍♂️

**Reply by jhirn** [reply]
got the magic from https://github.com/palkan/anyway_config/blob/a8d89abc72e036a675de01f9bac5f5f502d2c7a2/lib/anyway/rails/loaders/credentials.rb#L16

`ApplicationConfig.subclasses.each(&:new) unless defined?(Rails::Command::CredentialsCommand)`

**Comment by palkan** [comment]
I think, `ApplicationConfig.subclasses.each(&:new) unless defined?(Rails::Command::CredentialsCommand)` is good starting point.

We can try to incorporate something like this into a library to take into account all possible situations when we don't need to call this validation.

For Rails application, I'd suggest putting this into a server start callback:

```ruby
# application.rb
class MyApp < Rails::Application
  # ...
  server do
    ApplicationConfig.subclasses.each(&:new)
  end
end
```

**Reply by jhirn** [reply]
I never knew about the server block. TIL. 

I'm not sure if I want it to fail when starting a console but it feels like a good mechanism for fail fast startup checks. Will have to consider for that or things like runner. My initial thought is to always fail except when editing as the point is to add the thing making it fail. Part of me wishes edit didn't load application.rb at all but I understand it needs to for secrets config stuff. 

I really appreciate the discussion, suggestions and consideration of feature additions. I don't mean to sound as if I'm complaining or requesting feature. Just sharing experience as I'm converting a 6yr old app with a lot of curveballs. Hopefully the feedback helps you or others to use this amazing library. :)
